lib/sidekiq_unique_jobs/lock/until_and_while_executing.rb

```diff
@@ -23,8 +23,8 @@ class UntilAndWhileExecuting < BaseLock
       # @yield to the caller when given a block
       #
       def lock(origin: :client)
-        return lock_failed(origin: origin) unless (token = locksmith.lock)
-        return yield token if block_given?
+        token = locksmith.lock || lock_failed(origin: origin)
```

Comment:
The replace strategy should only be used for either the server or the client process if I remember correctly. 

Secondly, the strategy should run instead of the lock. Any return value from the strategy is coincidental and should not cause the job to execute which this implementation does. 

In other words, this is not the right solution.

Reply:
So I had a look now, you can get this information if you validate your worker.

```ruby
expect(MyWorker).to have_valid_sidekiq_options
```

```ruby
sidekiq_options: lock: :until_and_while_executing, on_conflict: { client: :replace, server: :reschedule }
```

Or if you rather raise and let the job be retried when it is able to get the lock:

```ruby
sidekiq_options: lock: :until_and_while_executing, on_conflict: { client: :replace, server: :raise }
```

Reply:
Sorry I messed up the formatting of the commit message. For us this is a regression caused by https://github.com/mhenrixon/sidekiq-unique-jobs/commit/8c8d54c8b9dea363a7d8b8aeaceb2e82966b8503. Before this change the test I added worked without an exception and behaved like we expected. We might have used a non-feature that wasn't intended to work. Since we are using `:until_and_while_executing` I don't think that the `server` `on_conflict` strategy will ever trigger. We are scheduling jobs on the client side to basically de-bounce events. For a group of events we only want one job to run X seconds after the last of the grouped events fired. This worked fine before and should never need a `server` `on_conflict` strategy because the jobs are scheduled to run in the future and should always only have one job per `lock_args` in the queue.

The exception I posted happens because `Middleware::Client#lock lock_instance.lock` now doesn't yield but returns the return value of `BaseLock#lock_failed` when the job was locked by strategy. This breaks the Sidekiq middleware interface because the middleware is supposed to either `yield` or return an `item` or `nil/false`. I believe the change restores the previous behavior if yielding inside `lock` when a strategy was successful:
https://github.com/mhenrixon/sidekiq-unique-jobs/blob/f4f4ce33124e17a252e9bf984fa0c05ec07cf342/lib/sidekiq_unique_jobs/lock/base_lock.rb#L45  
https://github.com/mhenrixon/sidekiq-unique-jobs/blob/f4f4ce33124e17a252e9bf984fa0c05ec07cf342/lib/sidekiq_unique_jobs/middleware/client.rb#L28

I just saw the current documentation of `lock_failed` which mentions it's supposed to return `void`, but I belive it still returns what it did before (`String, nil`) so the change might work for our use case by accident 😄 

Maybe the proper fix would be to pass the block into the `on_conflict` strategies and yield there if they were successful?

Reply:
Went through the code again and i now believe the return value of `OnConflict::Strategy#call` isn't documented correctly. It should have the same as `BaseLock#lock`. I think the change works (for our use case) because for `on_conflict: replace`, `lock_failed` basically behaves like `lock`:

https://github.com/mhenrixon/sidekiq-unique-jobs/blob/da53bb2fc8de71fec0aa4838b73263af5a03294f/lib/sidekiq_unique_jobs/lock/base_lock.rb#L109

https://github.com/mhenrixon/sidekiq-unique-jobs/blob/da53bb2fc8de71fec0aa4838b73263af5a03294f/lib/sidekiq_unique_jobs/lock/base_lock.rb#L119

https://github.com/mhenrixon/sidekiq-unique-jobs/blob/da53bb2fc8de71fec0aa4838b73263af5a03294f/lib/sidekiq_unique_jobs/on_conflict/replace.rb#L45

Which then calls `BaseLock#lock` again and returns `String, nil`.

WDYT about changing the return type of the strategies to `String, nil` and make all but `replace` return `nil` explicitly instead of implicitly like it's done for most of them currently?